libtock/internal/nonvolatile_storage_internal.c

```diff
@@ -1,31 +1,52 @@
 #include "internal/nonvolatile_storage.h"
 
-int nonvolatile_storage_internal_read_done_subscribe(subscribe_cb cb, void *userdata) {
-  return subscribe(DRIVER_NUM_NONVOLATILE_STORAGE, 0, cb, userdata);
+subscribe_return_t nonvolatile_storage_internal_read_done_subscribe(subscribe_cb cb, void *userdata) {
+  return subscribe2(DRIVER_NUM_NONVOLATILE_STORAGE, 0, cb, userdata);
 }
 
-int nonvolatile_storage_internal_write_done_subscribe(subscribe_cb cb, void *userdata) {
-  return subscribe(DRIVER_NUM_NONVOLATILE_STORAGE, 1, cb, userdata);
+subscribe_return_t nonvolatile_storage_internal_write_done_subscribe(subscribe_cb cb, void *userdata) {
+  return subscribe2(DRIVER_NUM_NONVOLATILE_STORAGE, 1, cb, userdata);
 }
 
-int nonvolatile_storage_internal_read_buffer(uint8_t* buffer, uint32_t len) {
-  return allow(DRIVER_NUM_NONVOLATILE_STORAGE, 0, (void*) buffer, len);
+allow_rw_return_t nonvolatile_storage_internal_read_buffer(uint8_t* buffer, uint32_t len) {
+  return allow_readwrite(DRIVER_NUM_NONVOLATILE_STORAGE, 0, (void*) buffer, len);
 }
 
-int nonvolatile_storage_internal_write_buffer(uint8_t* buffer, uint32_t len) {
-  return allow(DRIVER_NUM_NONVOLATILE_STORAGE, 1, (void*) buffer, len);
+allow_ro_return_t nonvolatile_storage_internal_write_buffer(uint8_t* buffer, uint32_t len) {
+  return allow_readonly(DRIVER_NUM_NONVOLATILE_STORAGE, 0, (void*) buffer, len);
 }
 
 int nonvolatile_storage_internal_get_number_bytes(void) {
-  return command(DRIVER_NUM_NONVOLATILE_STORAGE, 1, 0, 0);
+  syscall_return_t res = command2(DRIVER_NUM_NONVOLATILE_STORAGE, 1, 0, 0);
+  if (res.type == TOCK_SYSCALL_SUCCESS_U32) {
+    return res.data[0];
+  } else if (res.type == TOCK_SYSCALL_FAILURE) {
+    return tock_error_to_rcode(res.data[0]);
+  } else {
+    return TOCK_EBADRVAL;
+  }
 }
 
 int nonvolatile_storage_internal_read(uint32_t offset, uint32_t length) {
-  uint32_t arg0 = (length << 8) | 2;
-  return command(DRIVER_NUM_NONVOLATILE_STORAGE, (int) arg0, (int) offset, 0);
+  syscall_return_t res =
+    command2(DRIVER_NUM_NONVOLATILE_STORAGE, 2, (int) offset, (int) length);
```

Comment:
what is this change? I may have just missed a change in the kernel driver

Reply:
Yes. The ABI was changed slightly, as we had sufficient arguments available such that we didn't need to pack with the length and offset into a single register. Is there a good reason not to do this?

Reply:
oh I see. Nope that makes sense!